File: python/prism_sdk/autonomous_goal_worker.py
"""Execute scheduled goals through caller-owned task rehydration.

The goal ledger and scheduler intentionally do not retain task text.  This module closes the
runtime loop without weakening that boundary: a resolver briefly rehydrates a task from the
application's protected work queue, an executor performs one approved attempt, and the worker
settles only status, criterion/evaluator digests, and bounded failure metadata.  The resolver and
executor are never serialized, and the live execution result is available only to the initiating
caller through the in-memory batch object.
"""

from __future__ import annotations

from dataclasses import dataclass
from collections.abc import Callable, Mapping, Sequence
from typing import Any, Literal

from .authoring import content_digest
from .autonomous_goal_scheduler import (
    MAX_GOAL_SCHEDULE_SELECTED,
    AutonomousGoalSchedule,
    AutonomousGoalScheduleRow,
    AutonomousGoalScheduler,
    AutonomousGoalClaimResult,
)
from .goals import (
    GOAL_RETENTION,
    AutonomousGoalError,
    AutonomousGoalLedger,
    AutonomousGoalRecord,
    GoalStatus,
    goal_status_for_result,
    goal_task_digest,
)


GOAL_WORKER_SCHEMA = "bioprism-autonomous-goal-worker/0.1"
GOAL_WORKER_RETENTION = "metadata_only_goal_execution;task_and_execution_values_not_retained"
MAX_GOAL_WORKER_RUNS = MAX_GOAL_SCHEDULE_SELECTED
MAX_GOAL_WORKER_TASK_BYTES = 32_000
_SETTLEMENT_KEYS = frozenset({"evaluator_digest", "learning_state_digest", "progress_digest"})

WorkerRunStatus = Literal["completed", "paused", "blocked", "failed"]
GoalResolver = Callable[[AutonomousGoalRecord, AutonomousGoalScheduleRow], Mapping[str, Any]]
GoalExecutor = Callable[["AutonomousGoalExecutionRequest"], Any]


def _fail(message: str) -> None:
    raise AutonomousGoalError(f"autonomous goal worker {message}")


def _digest(value: Any) -> str:
    return content_digest(value)


def _task(value: Any) -> str:
    if not isinstance(value, str) or not value.strip() or "\x00" in value:
        _fail("resolved task must be a non-empty NUL-free string")
    if len(value.encode("utf-8")) > MAX_GOAL_WORKER_TASK_BYTES:
        _fail("resolved task exceeds its bounded input size")
    return value


def _status(value: Any) -> str:
    if isinstance(value, Mapping):
        value = value.get("status")
    else:
        value = getattr(value, "status", None)
    if not isinstance(value, str) or not value.strip() or "\x00" in value or len(value.encode("utf-8")) > 128:
        _fail("executor result status is outside its bounded contract")
    return value.strip()


def _field(value: Any, name: str, default: Any) -> Any:
    if isinstance(value, Mapping):
        return value.get(name, default)
    return getattr(value, name, default)


def _settlement_metadata(value: Any) -> dict[str, str | None]:
    raw = _field(value, "settlement_metadata", {})
    if raw is None:
        return {}
    if not isinstance(raw, Mapping):
        _fail("executor settlement_metadata must be a mapping")
    unknown = sorted(set(raw).difference(_SETTLEMENT_KEYS))
    if unknown:
        _fail("executor settlement_metadata contains unsupported fields: " + ", ".join(unknown))
    normalized: dict[str, str | None] = {}
    for name, item in raw.items():
        if item is not None and (not isinstance(item, str) or len(item) != 64 or any(char not in "0123456789abcdef" for char in item)):
            _fail(f"executor settlement_metadata.{name} must be a lowercase SHA-256 digest or None")
        normalized[name] = item
    return normalized


def _criterion_updates(value: Any) -> tuple[Mapping[str, Any], ...]:
    raw = _field(value, "criterion_updates", ())
    if raw is None:
        return ()
    if not isinstance(raw, Sequence) or isinstance(raw, (str, bytes, bytearray)) or len(raw) > 64:
        _fail("executor criterion_updates are outside their bounds")
    if any(not isinstance(item, Mapping) for item in raw):
        _fail("executor criterion_updates must contain mappings")
    return tuple(raw)


@dataclass(frozen=True, slots=True)
class AutonomousGoalExecutionRequest:
    """Transient rehydrated work passed to one caller-owned executor."""

    goal: AutonomousGoalRecord
    schedule_row: AutonomousGoalScheduleRow
    task: str
    parameters: Mapping[str, Any]
    schedule_digest: str

    def metadata(self) -> dict[str, Any]:
        return {
            "goal_id": self.goal.goal_id,
            "domain": self.goal.domain,
            "attempt": self.goal.attempt,
            "revision": self.goal.revision,
            "schedule_digest": self.schedule_digest,
        }


@dataclass(frozen=True, slots=True)
class AutonomousGoalWorkerRun:
    goal_id: str
    domain: str
    attempt: int
    execution_status: WorkerRunStatus
    goal_status: GoalStatus
    outcome_digest: str
    schedule_digest: str
    claim_digest: str
    error_class: str | None = None
    error_digest: str | None = None
    live_result: Any = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "goal_id": self.goal_id,
            "domain": self.domain,
            "attempt": self.attempt,
            "execution_status": self.execution_status,
            "goal_status": self.goal_status,
            "outcome_digest": self.outcome_digest,
            "schedule_digest": self.schedule_digest,
            "claim_digest": self.claim_digest,
            "error_class": self.error_class,
            "error_digest": self.error_digest,
        }


@dataclass(frozen=True, slots=True)
class AutonomousGoalWorkerBatch:
    schedule: AutonomousGoalSchedule
    claim: AutonomousGoalClaimResult | None
    runs: tuple[AutonomousGoalWorkerRun, ...]
    worker_digest: str

    @property
    def live_results(self) -> tuple[Any, ...]:
        return tuple(run.live_result for run in self.runs)

    def to_dict(self) -> dict[str, Any]:
        claim = None if self.claim is None else self.claim.to_dict()
        rows = [run.to_dict() for run in self.runs]
        body = {
            "schema": GOAL_WORKER_SCHEMA,
            "schedule": self.schedule.to_dict(),
            "claim": claim,
            "runs": rows,
            "counts": {
                "selected": len(self.schedule.selected_goal_ids),
                "claimed": 0 if self.claim is None else len(self.claim.claims),
                "settled": len(rows),
                "completed": sum(run.goal_status == "completed" for run in self.runs),
                "paused": sum(run.goal_status == "paused" for run in self.runs),
                "blocked": sum(run.goal_status == "blocked" for run in self.runs),
                "failed": sum(run.goal_status == "failed" for run in self.runs),
            },
            "worker_digest": self.worker_digest,
            "retention": GOAL_WORKER_RETENTION,
            "goal_retention": GOAL_RETENTION,
            "secret_material": "never_returned",
        }
        return body


class AutonomousGoalWorker:
    """Schedule, claim, rehydrate, execute, and settle a bounded goal batch.

    ``resolver`` is the only place task text enters the worker.  It must return a transient
    mapping containing ``task`` and may return ``parameters`` for the executor.  Neither is
    copied into the ledger, schedule, claim receipt, worker digest, or public ``to_dict`` view.
    ``executor`` may return a status-bearing object/mapping plus optional ``criterion_updates``
    and digest-only ``settlement_metadata``.
    """

    def __init__(
        self,
        ledger: AutonomousGoalLedger,
        *,
        resolver: GoalResolver,
        executor: GoalExecutor,
        scheduler: AutonomousGoalScheduler | None = None,
    ) -> None:
        if not isinstance(ledger, AutonomousGoalLedger):
            _fail("ledger must be an AutonomousGoalLedger")
        if not callable(resolver):
            _fail("resolver must be callable")
        if not callable(executor):
            _fail("executor must be callable")
        if scheduler is not None and not isinstance(scheduler, AutonomousGoalScheduler):
            _fail("scheduler must be an AutonomousGoalScheduler")
        self.ledger = ledger
        self.resolver = resolver
        self.executor = executor
        self.scheduler = scheduler or AutonomousGoalScheduler()

    def run(self, *, schedule_options: Mapping[str, Any] | None = None) -> AutonomousGoalWorkerBatch:
        if schedule_options is not None and not isinstance(schedule_options, Mapping):
            _fail("schedule_options must be a mapping or None")
        options = {} if schedule_options is None else dict(schedule_options)
        # The ledger intentionally caps one bounded listing at 512 rows.  The scheduler's
        # admission cap is smaller, so this is enough to make one worker pass deterministic.
        goals = self.ledger.list(limit=512)
        schedule = self.scheduler.plan(goals, options)
        rows = {row.goal_id: row for row in schedule.rows if row.decision == "admit"}
        prepared: dict[str, AutonomousGoalExecutionRequest] = {}
        for goal_id in schedule.selected_goal_ids:
            goal = self.ledger.get(goal_id)
            row = rows.get(goal_id)
            if goal is None or row is None:
                _fail(f"schedule admission disappeared for goal {goal_id}")
            resolved = self.resolver(goal, row)
            if not isinstance(resolved, Mapping):
                _fail(f"resolver returned a non-mapping for goal {goal_id}")
            task = _task(resolved.get("task"))
            resolved_domain = resolved.get("domain", goal.domain)
            if resolved_domain != goal.domain:
                _fail(f"resolver domain does not match goal {goal_id}")
            parameters = resolved.get("parameters", {})
            if not isinstance(parameters, Mapping):
                _fail(f"resolver parameters must be a mapping for goal {goal_id}")
            prepared[goal_id] = AutonomousGoalExecutionRequest(
                goal=goal,
                schedule_row=row,
                task=task,
                parameters=dict(parameters),
                schedule_digest=schedule.schedule_digest,
            )
        claim = self.scheduler.claim(
            self.ledger,
            schedule,
            now_ns=options.get("now_ns") if isinstance(options.get("now_ns"), int) else None,
        ) if schedule.selected_goal_ids else None
        runs: list[AutonomousGoalWorkerRun] = []
        if claim is not None:
            claim_by_id = {item.goal_id: item for item in claim.claims}
            for goal_id in schedule.selected_goal_ids:
                claim_row = claim_by_id[goal_id]
                request = prepared[goal_id]
                current = self.ledger.get(goal_id)
                if current is None or current.status != "running" or current.revision != claim_row.running_revision:
                    _fail(f"claimed goal {goal_id} changed before execution")
                try:
                    live_result = self.executor(request)
                    result_status = _status(live_result)
                    updated = self._settle(current, result_status, live_result)
                    outcome_digest = _digest({"goal_id": goal_id, "attempt": current.attempt, "result_status": result_status})
                    runs.append(
                        AutonomousGoalWorkerRun(
                            goal_id=goal_id,
                            domain=current.domain,
                            attempt=current.attempt,
                            execution_status=updated.status,
                            goal_status=updated.status,
                            outcome_digest=outcome_digest,
                            schedule_digest=schedule.schedule_digest,
                            claim_digest=claim.claim_digest,
                            live_result=live_result,
                        )
                    )
                except Exception as error:
                    error_class = type(error).__name__
                    outcome_digest = _digest({"goal_id": goal_id, "attempt": current.attempt, "result_status": f"exception:{error_class}"})
                    try:
                        updated = self.ledger.transition(
                            goal_id,
                            "failed",
                            expected_revision=current.revision,
                            blockers=(f"exception:{error_class}",),
                            next_action_digest=goal_task_digest("goal-retry"),
                            outcome_digest=outcome_digest,
                        )
                    except AutonomousGoalError as transition_error:
                        raise AutonomousGoalError(f"goal {goal_id} failed without a durable failure transition") from transition_error
                    runs.append(
                        AutonomousGoalWorkerRun(
                            goal_id=goal_id,
                            domain=current.domain,
                            attempt=current.attempt,
                            execution_status="failed",
                            goal_status=updated.status,
                            outcome_digest=outcome_digest,
                            schedule_digest=schedule.schedule_digest,
                            claim_digest=claim.claim_digest,
                            error_class=error_class,
                            error_digest=_digest({"error_class": error_class}),
                        )
                    )
        body = {
            "schema": GOAL_WORKER_SCHEMA,
            "schedule_digest": schedule.schedule_digest,
            "claim_digest": None if claim is None else claim.claim_digest,
            "runs": [run.to_dict() for run in runs],
            "retention": GOAL_WORKER_RETENTION,
            "goal_retention": GOAL_RETENTION,
            "secret_material": "never_returned",
        }
        return AutonomousGoalWorkerBatch(
            schedule=schedule,
            claim=claim,
            runs=tuple(runs),
            worker_digest=_digest(body),
        )

    def _settle(self, current: AutonomousGoalRecord, result_status: str, result: Any) -> AutonomousGoalRecord:
        outcome_digest = _digest({"goal_id": current.goal_id, "attempt": current.attempt, "result_status": result_status})
        updates = _criterion_updates(result)
        settled = current
        if updates:
            settled = self.ledger.update_criteria(current.goal_id, updates, expected_revision=current.revision)
        metadata = _settlement_metadata(result)
        target = goal_status_for_result(result_status, criteria_complete=settled.required_criteria_complete)
        transition_metadata = {key: value for key, value in metadata.items() if value is not None}
        return self.ledger.transition(
            current.goal_id,
            target,
            expected_revision=settled.revision,
            blockers=(() if target == "completed" else (f"result:{result_status}",)),
            next_action_digest=(None if target == "completed" else goal_task_digest(f"goal-next:{result_status}")),
            outcome_digest=outcome_digest,
            **transition_metadata,
        )


__all__ = [
    "GOAL_WORKER_RETENTION",
    "GOAL_WORKER_SCHEMA",
    "MAX_GOAL_WORKER_RUNS",
    "MAX_GOAL_WORKER_TASK_BYTES",
    "AutonomousGoalExecutionRequest",
    "AutonomousGoalWorker",
    "AutonomousGoalWorkerBatch",
    "AutonomousGoalWorkerRun",
]

File: typescript/src/autonomous-goal-worker.ts
import { ArgumentError, isObject } from "./errors.js";
import {
  AUTONOMOUS_GOAL_RETENTION,
  goalStatusForResult,
  goalTaskDigest,
  InMemoryAutonomousGoalLedger,
  type AutonomousGoalRecord,
  type AutonomousGoalStatus,
} from "./autonomous-goals.js";
import {
  AUTONOMOUS_GOAL_SCHEDULE_MAX_SELECTED,
  AutonomousGoalScheduler,
  type AutonomousGoalSchedule,
  type AutonomousGoalScheduleRow,
  type AutonomousGoalClaimResult,
} from "./autonomous-goal-scheduler.js";
import { AutonomousGoalWorkerJournal } from "./autonomous-goal-worker-journal.js";
import { digestJsonSync } from "./tooling.js";
import type { JsonObject } from "./types.js";

/** The transient rehydration bridge between a metadata-only goal and a caller-owned executor. */
export const AUTONOMOUS_GOAL_WORKER_SCHEMA = "bioprism-autonomous-goal-worker/0.1" as const;
export const AUTONOMOUS_GOAL_WORKER_RETENTION = "metadata_only_goal_execution;task_and_execution_values_not_retained" as const;
export const AUTONOMOUS_GOAL_WORKER_MAX_RUNS = AUTONOMOUS_GOAL_SCHEDULE_MAX_SELECTED;
export const AUTONOMOUS_GOAL_WORKER_MAX_TASK_BYTES = 32_000;

export type AutonomousGoalWorkerRunStatus = "completed" | "paused" | "blocked" | "failed";

export interface AutonomousGoalExecutionRequest {
  goal: AutonomousGoalRecord;
  schedule_row: AutonomousGoalScheduleRow;
  task: string;
  parameters: JsonObject;
  schedule_digest: string;
  task_digest: string;
  execution_binding_digest: string;
}

export interface AutonomousGoalWorkerResolution extends JsonObject {
  task: string;
  domain?: string;
  parameters?: JsonObject;
}

export interface AutonomousGoalWorkerOutcome extends JsonObject {
  status: string;
  criterion_updates?: JsonObject[];
  settlement_metadata?: {
    evaluator_digest?: string | null;
    learning_state_digest?: string | null;
    progress_digest?: string | null;
  };
}

export type AutonomousGoalResolver = (goal: AutonomousGoalRecord, row: AutonomousGoalScheduleRow) => AutonomousGoalWorkerResolution | Promise<AutonomousGoalWorkerResolution>;
export type AutonomousGoalExecutor = (request: AutonomousGoalExecutionRequest) => unknown | Promise<unknown>;

export interface AutonomousGoalWorkerRun extends JsonObject {
  goal_id: string;
  domain: string;
  attempt: number;
  execution_status: AutonomousGoalWorkerRunStatus;
  goal_status: AutonomousGoalStatus;
  outcome_digest: string;
  schedule_digest: string;
  claim_digest: string;
  error_class: string | null;
  error_digest: string | null;
}

type LiveGoalWorkerRun = AutonomousGoalWorkerRun & { live_result?: any };

export interface AutonomousGoalWorkerBatchJSON extends JsonObject {
  schema: typeof AUTONOMOUS_GOAL_WORKER_SCHEMA;
  schedule: AutonomousGoalSchedule;
  claim: AutonomousGoalClaimResult | null;
  runs: AutonomousGoalWorkerRun[];
  counts: {
    selected: number;
    claimed: number;
    settled: number;
    completed: number;
    paused: number;
    blocked: number;
    failed: number;
  };
  worker_digest: string;
  retention: typeof AUTONOMOUS_GOAL_WORKER_RETENTION;
  goal_retention: typeof AUTONOMOUS_GOAL_RETENTION;
  secret_material: "never_returned";
}

export class AutonomousGoalWorkerBatch {
  constructor(
    readonly schedule: AutonomousGoalSchedule,
    readonly claim: AutonomousGoalClaimResult | null,
    readonly runs: readonly LiveGoalWorkerRun[],
    readonly worker_digest: string,
  ) {}

  get live_results(): unknown[] {
    return this.runs.map((run) => run.live_result);
  }

  toJSON(): AutonomousGoalWorkerBatchJSON {
    const runs = this.runs.map((run) => {
      const { live_result: _liveResult, ...metadata } = run;
      return metadata;
    });
    return {
      schema: AUTONOMOUS_GOAL_WORKER_SCHEMA,
      schedule: this.schedule,
      claim: this.claim,
      runs,
      counts: {
        selected: this.schedule.selected_goal_ids.length,
        claimed: this.claim?.claims.length ?? 0,
        settled: runs.length,
        completed: runs.filter((run) => run.goal_status === "completed").length,
        paused: runs.filter((run) => run.goal_status === "paused").length,
        blocked: runs.filter((run) => run.goal_status === "blocked").length,
        failed: runs.filter((run) => run.goal_status === "failed").length,
      },
      worker_digest: this.worker_digest,
      retention: AUTONOMOUS_GOAL_WORKER_RETENTION,
      goal_retention: AUTONOMOUS_GOAL_RETENTION,
      secret_material: "never_returned",
    };
  }
}

function fail(message: string): never {
  throw new ArgumentError(`autonomous goal worker ${message}`);
}

function digest(value: unknown): string {
  return digestJsonSync(value);
}

function task(value: unknown): string {
  if (typeof value !== "string" || !value.trim() || value.includes("\u0000") || new TextEncoder().encode(value).byteLength > AUTONOMOUS_GOAL_WORKER_MAX_TASK_BYTES) fail("resolved task is outside its bounded contract");
  return value;
}

function status(value: unknown): string {
  const candidate = isObject(value) ? value.status : (value as { status?: unknown } | null)?.status;
  if (typeof candidate !== "string" || !candidate.trim() || candidate.includes("\u0000") || new TextEncoder().encode(candidate).byteLength > 128) fail("executor result status is outside its bounded contract");
  return candidate.trim();
}

function field(value: unknown, name: string, fallback: unknown): unknown {
  return isObject(value) && name in value ? value[name] : (value as Record<string, unknown> | null)?.[name] ?? fallback;
}

function criterionUpdates(value: unknown): JsonObject[] {
  const raw = field(value, "criterion_updates", []);
  if (raw === null || raw === undefined) return [];
  if (!Array.isArray(raw) || raw.length > 64 || raw.some((item) => !isObject(item))) fail("executor criterion_updates are outside their bounds");
  return raw as JsonObject[];
}

function settlementMetadata(value: unknown): Record<string, string | null> {
  const raw = field(value, "settlement_metadata", {});
  if (raw === null || raw === undefined) return {};
  if (!isObject(raw)) fail("executor settlement_metadata must be an object");
  const allowed = new Set(["evaluator_digest", "learning_state_digest", "progress_digest"]);
  if (Object.keys(raw).some((key) => !allowed.has(key))) fail("executor settlement_metadata contains unsupported fields");
  const normalized: Record<string, string | null> = {};
  for (const [key, item] of Object.entries(raw)) {
    if (item !== null && (typeof item !== "string" || !/^[0-9a-f]{64}$/.test(item))) fail(`executor settlement_metadata.${key} must be a lowercase SHA-256 digest or null`);
    normalized[key] = item as string | null;
  }
  return normalized;
}

export class AutonomousGoalWorker {
  readonly ledger: InMemoryAutonomousGoalLedger;
  readonly resolver: AutonomousGoalResolver;
  readonly executor: AutonomousGoalExecutor;
  readonly scheduler: AutonomousGoalScheduler;
  readonly journal: AutonomousGoalWorkerJournal | undefined;

  constructor(options: { ledger: InMemoryAutonomousGoalLedger; resolver: AutonomousGoalResolver; executor: AutonomousGoalExecutor; scheduler?: AutonomousGoalScheduler; journal?: AutonomousGoalWorkerJournal }) {
    if (!(options?.ledger instanceof InMemoryAutonomousGoalLedger)) fail("ledger must be an InMemoryAutonomousGoalLedger");
    if (typeof options.resolver !== "function") fail("resolver must be callable");
    if (typeof options.executor !== "function") fail("executor must be callable");
    if (options.scheduler !== undefined && !(options.scheduler instanceof AutonomousGoalScheduler)) fail("scheduler must be an AutonomousGoalScheduler");
    if (options.journal !== undefined && !(options.journal instanceof AutonomousGoalWorkerJournal)) fail("journal must be an AutonomousGoalWorkerJournal");
    this.ledger = options.ledger;
    this.resolver = options.resolver;
    this.executor = options.executor;
    this.scheduler = options.scheduler ?? new AutonomousGoalScheduler();
    this.journal = options.journal;
  }

  async run(options: { schedule_options?: Record<string, unknown>; batch_id?: string } = {}): Promise<AutonomousGoalWorkerBatch> {
    if (options.schedule_options !== undefined && !isObject(options.schedule_options)) fail("schedule_options must be an object");
    if (this.journal !== undefined && (typeof options.batch_id !== "string" || !options.batch_id.trim() || options.batch_id.includes("\u0000") || new TextEncoder().encode(options.batch_id).byteLength > 256)) fail("batch_id is required and bounded when a journal is configured");
    const scheduleOptions = options.schedule_options ?? {};
    // The ledger intentionally caps one bounded listing at 512 rows.  The scheduler's
    // admission cap is smaller, so this is enough to make one worker pass deterministic.
    const schedule = this.scheduler.plan(this.ledger.list({ limit: 512 }), scheduleOptions);
    const rows = new Map(schedule.rows.filter((row) => row.decision === "admit").map((row) => [row.goal_id, row]));
    const prepared = new Map<string, AutonomousGoalExecutionRequest>();
    for (const goalId of schedule.selected_goal_ids) {
      const goal = this.ledger.get(goalId);
      const row = rows.get(goalId);
      if (!goal || !row) fail(`schedule admission disappeared for goal ${goalId}`);
      this.journal?.assertNoActive(goalId);
      const resolved = await this.resolver(goal, row);
      if (!isObject(resolved)) fail(`resolver returned a non-object for goal ${goalId}`);
      const resolvedDomain = resolved.domain ?? goal.domain;
      if (resolvedDomain !== goal.domain) fail(`resolver domain does not match goal ${goalId}`);
      const parameters = resolved.parameters ?? {};
      if (!isObject(parameters)) fail(`resolver parameters must be an object for goal ${goalId}`);
      const resolvedTask = task(resolved.task);
      if (goalTaskDigest(resolvedTask) !== goal.task_digest) fail(`resolver task digest does not match goal ${goalId}`);
      const clonedParameters = structuredClone(parameters);
      prepared.set(goalId, { goal, schedule_row: row, task: resolvedTask, parameters: clonedParameters, schedule_digest: schedule.schedule_digest, task_digest: goal.task_digest, execution_binding_digest: digest({ parameters: clonedParameters }) });
    }
    if (this.journal !== undefined && options.batch_id !== undefined) {
      for (const request of prepared.values()) this.journal.record({ batch_id: options.batch_id, goal_id: request.goal.goal_id, phase: "prepared", attempt: request.goal.attempt, revision: request.goal.revision, schedule_digest: schedule.schedule_digest, task_digest: request.task_digest, execution_binding_digest: request.execution_binding_digest });
    }
    const claim = schedule.selected_goal_ids.length === 0 ? null : this.scheduler.claim(this.ledger, schedule, { now_ns: typeof scheduleOptions.now_ns === "number" ? scheduleOptions.now_ns : undefined });
    const runs: LiveGoalWorkerRun[] = [];
    if (claim) {
      const claimById = new Map(claim.claims.map((item) => [item.goal_id, item]));
      if (this.journal !== undefined && options.batch_id !== undefined) {
        for (const item of claim.claims) {
          const current = this.ledger.get(item.goal_id);
          if (!current) fail(`claimed goal ${item.goal_id} disappeared before journaling`);
          const request = prepared.get(item.goal_id);
          if (!request) fail(`prepared request for goal ${item.goal_id} disappeared before journaling`);
          this.journal.record({ batch_id: options.batch_id, goal_id: item.goal_id, phase: "claimed", attempt: current.attempt, revision: current.revision, schedule_digest: schedule.schedule_digest, claim_digest: claim.claim_digest, task_digest: request.task_digest, execution_binding_digest: request.execution_binding_digest });
        }
      }
      for (const goalId of schedule.selected_goal_ids) {
        const claimRow = claimById.get(goalId)!;
        const request = prepared.get(goalId)!;
        const current = this.ledger.get(goalId);
        if (!current || current.status !== "running" || current.revision !== claimRow.running_revision) fail(`claimed goal ${goalId} changed before execution`);
        try {
          if (this.journal !== undefined && options.batch_id !== undefined) this.journal.record({ batch_id: options.batch_id, goal_id: goalId, phase: "dispatch_started", attempt: current.attempt, revision: current.revision, schedule_digest: schedule.schedule_digest, claim_digest: claim.claim_digest, task_digest: request.task_digest, execution_binding_digest: request.execution_binding_digest });
          const liveResult = await this.executor(request);
          const resultStatus = status(liveResult);
          const updated = this.settle(current, resultStatus, liveResult);
          const outcomeDigest = digest({ goal_id: goalId, attempt: current.attempt, result_status: resultStatus });
          if (this.journal !== undefined && options.batch_id !== undefined) this.journal.record({ batch_id: options.batch_id, goal_id: goalId, phase: "settled", attempt: current.attempt, revision: updated.revision, schedule_digest: schedule.schedule_digest, claim_digest: claim.claim_digest, outcome_digest: outcomeDigest, task_digest: request.task_digest, execution_binding_digest: request.execution_binding_digest });
          runs.push({ goal_id: goalId, domain: current.domain, attempt: current.attempt, execution_status: updated.status as AutonomousGoalWorkerRunStatus, goal_status: updated.status, outcome_digest: outcomeDigest, schedule_digest: schedule.schedule_digest, claim_digest: claim.claim_digest, error_class: null, error_digest: null, live_result: liveResult as any });
        } catch (error) {
          const errorClass = error instanceof Error ? error.constructor.name : "UnknownError";
          const outcomeDigest = digest({ goal_id: goalId, attempt: current.attempt, result_status: `exception:${errorClass}` });
          let updated;
          try {
            updated = this.ledger.transition(goalId, "failed", { expected_revision: current.revision, blockers: [`exception:${errorClass}`], next_action_digest: goalTaskDigest("goal-retry"), outcome_digest: outcomeDigest });
          } catch (transitionError) {
            const wrapped = new ArgumentError(`goal ${goalId} failed without a durable failure transition`);
            (wrapped as Error & { cause?: unknown }).cause = transitionError;
            throw wrapped;
          }
          if (this.journal !== undefined && options.batch_id !== undefined) this.journal.record({ batch_id: options.batch_id, goal_id: goalId, phase: "failed", attempt: current.attempt, revision: updated.revision, schedule_digest: schedule.schedule_digest, claim_digest: claim.claim_digest, outcome_digest: outcomeDigest, error_digest: digest({ error_class: errorClass }), task_digest: request.task_digest, execution_binding_digest: request.execution_binding_digest });
          runs.push({ goal_id: goalId, domain: current.domain, attempt: current.attempt, execution_status: "failed", goal_status: updated.status, outcome_digest: outcomeDigest, schedule_digest: schedule.schedule_digest, claim_digest: claim.claim_digest, error_class: errorClass, error_digest: digest({ error_class: errorClass }) });
        }
      }
    }
    const metadataRuns = runs.map(({ live_result: _liveResult, ...metadata }) => metadata);
    const workerDigest = digest({ schema: AUTONOMOUS_GOAL_WORKER_SCHEMA, schedule_digest: schedule.schedule_digest, claim_digest: claim?.claim_digest ?? null, runs: metadataRuns, retention: AUTONOMOUS_GOAL_WORKER_RETENTION, goal_retention: AUTONOMOUS_GOAL_RETENTION, secret_material: "never_returned" });
    return new AutonomousGoalWorkerBatch(schedule, claim, runs, workerDigest);
  }

  private settle(current: AutonomousGoalRecord, resultStatus: string, result: unknown): AutonomousGoalRecord {
    const outcomeDigest = digest({ goal_id: current.goal_id, attempt: current.attempt, result_status: resultStatus });
    const updates = criterionUpdates(result);
    const settled = updates.length === 0 ? current : this.ledger.updateCriteria(current.goal_id, updates, { expected_revision: current.revision });
    const criteriaComplete = settled.criteria.every((criterion) => !criterion.required || ["satisfied", "waived"].includes(criterion.status));
    const target = goalStatusForResult(resultStatus, criteriaComplete);
    const metadata = settlementMetadata(result);
    return this.ledger.transition(current.goal_id, target, { expected_revision: settled.revision, blockers: target === "completed" ? [] : [`result:${resultStatus}`], next_action_digest: target === "completed" ? null : goalTaskDigest(`goal-next:${resultStatus}`), outcome_digest: outcomeDigest, ...metadata });
  }
}
